Support PHP 7.2

### Summary

#### What's the PHP version?
- Hello~ My PHP version is `PHP 7.2.18` in CentOS 7.2. I haven't found it in the doc.
- When I build it, it replies to some error.

### Additional details
- The error info like this:
```
/data/alpha/rust/php-ext-wasm-0.5.0/src/wasm.cc: In function ‘void zim_WasmArrayBuffer_getByteLength(zend_execute_data*, zval*)’:
/data/alpha/rust/php-ext-wasm-0.5.0/src/wasm.cc:139:32: error: ‘ZEND_PARSE_PARAMETERS_NONE’ was not declared in this scope
     ZEND_PARSE_PARAMETERS_NONE();
                                ^
/data/alpha/rust/php-ext-wasm-0.5.0/src/wasm.cc: In function ‘void zif_wasm_compile(zend_execute_data*, zval*)’:
/data/alpha/rust/php-ext-wasm-0.5.0/src/wasm.cc:505:13: error: ‘zend_register_persistent_resource_ex’ was not declared in this scope
             );
             ^
/data/alpha/rust/php-ext-wasm-0.5.0/src/wasm.cc: In function ‘void zif_wasm_module_clean_up_persistent_resources(zend_execute_data*, zval*)’:
/data/alpha/rust/php-ext-wasm-0.5.0/src/wasm.cc:571:32: error: ‘ZEND_PARSE_PARAMETERS_NONE’ was not declared in this scope
     ZEND_PARSE_PARAMETERS_NONE();
                                ^
In file included from /www/server/php/72/include/php/Zend/zend.h:29:0,
                 from /www/server/php/72/include/php/main/php.h:35,
                 from /data/alpha/rust/php-ext-wasm-0.5.0/src/wasm.hh:23,
                 from /data/alpha/rust/php-ext-wasm-0.5.0/src/wasm.cc:19:
/data/alpha/rust/php-ext-wasm-0.5.0/src/wasm.cc: At global scope:
/data/alpha/rust/php-ext-wasm-0.5.0/src/wasm.cc:1062:100: error: ‘_IS_NUMBER’ was not declared in this scope
 ZEND_BEGIN_ARG_WITH_RETURN_TYPE_INFO_EX(arginfo_wasm_invoke_function, ZEND_RETURN_VALUE, ARITY(3), _IS_NUMBER, NULLABLE)
                                                                                                    ^
/www/server/php/72/include/php/Zend/zend_types.h:143:5: note: in definition of macro ‘ZEND_TYPE_ENCODE’
  (((code) << Z_L(1)) | ((allow_null) ? Z_L(0x1) : Z_L(0x0)))
     ^
/data/alpha/rust/php-ext-wasm-0.5.0/src/wasm.cc:1062:1: note: in expansion of macro ‘ZEND_BEGIN_ARG_WITH_RETURN_TYPE_INFO_EX’
 ZEND_BEGIN_ARG_WITH_RETURN_TYPE_INFO_EX(arginfo_wasm_invoke_function, ZEND_RETURN_VALUE, ARITY(3), _IS_NUMBER, NULLABLE)
 ^
/data/alpha/rust/php-ext-wasm-0.5.0/src/wasm.cc: In function ‘void zif_wasm_get_last_error(zend_execute_data*, zval*)’:
/data/alpha/rust/php-ext-wasm-0.5.0/src/wasm.cc:1475:32: error: ‘ZEND_PARSE_PARAMETERS_NONE’ was not declared in this scope
     ZEND_PARSE_PARAMETERS_NONE();
                                ^
/data/alpha/rust/php-ext-wasm-0.5.0/src/wasm.cc: In function ‘void zif_WasmTypedArray_get_offset(zend_execute_data*, zval*)’:
/data/alpha/rust/php-ext-wasm-0.5.0/src/wasm.cc:1724:32: error: ‘ZEND_PARSE_PARAMETERS_NONE’ was not declared in this scope
     ZEND_PARSE_PARAMETERS_NONE();
                                ^
/data/alpha/rust/php-ext-wasm-0.5.0/src/wasm.cc: In function ‘void zif_WasmTypedArray_get_length(zend_execute_data*, zval*)’:
/data/alpha/rust/php-ext-wasm-0.5.0/src/wasm.cc:1751:32: error: ‘ZEND_PARSE_PARAMETERS_NONE’ was not declared in this scope
     ZEND_PARSE_PARAMETERS_NONE();
                                ^
In file included from /www/server/php/72/include/php/Zend/zend.h:29:0,
                 from /www/server/php/72/include/php/main/php.h:35,
                 from /data/alpha/rust/php-ext-wasm-0.5.0/src/wasm.hh:23,
                 from /data/alpha/rust/php-ext-wasm-0.5.0/src/wasm.cc:19:
/data/alpha/rust/php-ext-wasm-0.5.0/src/wasm.cc: At global scope:
/data/alpha/rust/php-ext-wasm-0.5.0/src/wasm.cc:1762:105: error: ‘_IS_NUMBER’ was not declared in this scope
 ZEND_BEGIN_ARG_WITH_RETURN_TYPE_INFO_EX(arginfo_wasmtypedarray_offset_get, ZEND_RETURN_VALUE, ARITY(1), _IS_NUMBER, NOT_NULLABLE)
                                                                                                         ^
/www/server/php/72/include/php/Zend/zend_types.h:143:5: note: in definition of macro ‘ZEND_TYPE_ENCODE’
  (((code) << Z_L(1)) | ((allow_null) ? Z_L(0x1) : Z_L(0x0)))
     ^
/data/alpha/rust/php-ext-wasm-0.5.0/src/wasm.cc:1762:1: note: in expansion of macro ‘ZEND_BEGIN_ARG_WITH_RETURN_TYPE_INFO_EX’
 ZEND_BEGIN_ARG_WITH_RETURN_TYPE_INFO_EX(arginfo_wasmtypedarray_offset_get, ZEND_RETURN_VALUE, ARITY(1), _IS_NUMBER, NOT_NULLABLE)
 ^
/data/alpha/rust/php-ext-wasm-0.5.0/src/wasm.cc: In function ‘int zm_startup_wasm(int, int)’:
/data/alpha/rust/php-ext-wasm-0.5.0/src/wasm.cc:2104:65: error: ‘ZEND_ACC_IMMUTABLE’ was not declared in this scope
     wasm_array_buffer_class_entry->ce_flags |= ZEND_ACC_FINAL | ZEND_ACC_IMMUTABLE;
                                                                 ^
make: *** [wasm.lo] Error 1
error: Recipe `build` failed with exit code 2
```

--- a/.circleci/config.yml
+++ b/.circleci/config.yml
@@ -1,65 +1,74 @@
 version: 2.1
 
+build_and_test_steps: &build_and_test_steps
+  # Update the project.
+  - checkout
+
+  # Install Rust.
+  - run:
+      name: Install Rust
+      command: |
+        test -d /usr/local/cargo || curl https://sh.rustup.rs -sSf | sh -s -- -y
+
+  # Install `just` used to manage the project.
+  - run:
+      name: Install just
+      command: |
+        export PATH="$HOME/.cargo/bin:$PATH"
+        test -f $HOME/.cargo/bin/just || cargo install just
+
+  # Install Composer.
+  - run:
+      name: Install Composer
+      command: |
+        php -r "copy('https://getcomposer.org/installer', 'composer-setup.php');"
+        php composer-setup.php
+        php -r "unlink('composer-setup.php');"
+        mv composer.phar composer
+
+  # Compile and install the PHP extension.
+  # `just build` isn't used because there is no some issue with
+  # `--with-pic` I don't understand…
+  - run:
+      name: Compile and install the PHP extension
+      command: |
+        export PATH="$HOME/.cargo/bin:$PATH"
+        export CXX="gcc"
+        EXTENSION="$HOME/project/src"
+        cd $EXTENSION
+        PHP_PREFIX=$(php-config --prefix)
+        PHP_PREFIX_BIN=$PHP_PREFIX/bin
+        $PHP_PREFIX_BIN/phpize --clean
+        $PHP_PREFIX_BIN/phpize
+        ./configure --with-php-config=$PHP_PREFIX_BIN/php-config
+        /bin/bash $EXTENSION/libtool --mode=compile $CXX -I. -I$EXTENSION -DPHP_ATOM_INC -I$EXTENSION/include -I$EXTENSION/main -I$EXTENSION -I$PHP_PREFIX/include/php -I$PHP_PREFIX/include/php/main -I$PHP_PREFIX/include/php/TSRM -I$PHP_PREFIX/include/php/Zend -I$PHP_PREFIX/include/php/ext -I$PHP_PREFIX/include/php/ext/date/lib -DHAVE_CONFIG_H -c $EXTENSION/wasm.cc -o wasm.lo -fPIC
+        $CXX -I. -I$EXTENSION -DPHP_ATOM_INC -I$EXTENSION/include -I$EXTENSION/main -I$EXTENSION -I$PHP_PREFIX/include/php -I$PHP_PREFIX/include/php/main -I$PHP_PREFIX/include/php/TSRM -I$PHP_PREFIX/include/php/Zend -I$PHP_PREFIX/include/php/ext -I$PHP_PREFIX/include/php/ext/date/lib -DHAVE_CONFIG_H -c $EXTENSION/wasm.cc  -DPIC -o .libs/wasm.o -fPIC
+        /bin/bash $EXTENSION/libtool --mode=link cc -DPHP_ATOM_INC -I$EXTENSION/include -I$EXTENSION/main -I$EXTENSION -I$PHP_PREFIX/include/php -I$PHP_PREFIX/include/php/main -I$PHP_PREFIX/include/php/TSRM -I$PHP_PREFIX/include/php/Zend -I$PHP_PREFIX/include/php/ext -I$PHP_PREFIX/include/php/ext/date/lib  -DHAVE_CONFIG_H  -g -O2    -o wasm.la -export-dynamic -avoid-version -prefer-pic -module -rpath $EXTENSION/modules  wasm.lo -Wl,-rpath,$EXTENSION/. -L$EXTENSION/. -lwasmer_runtime_c_api -fPIC
+        cc -shared  .libs/wasm.o  -L$EXTENSION/. -lwasmer_runtime_c_api  -Wl,-rpath -Wl,$EXTENSION/. -Wl,-soname -Wl,wasm.so -o .libs/wasm.so -fPIC
+        sudo make install-modules
+
+  # Run the extension test suites.
+  - run:
+      name: Test the extension
+      command: |
+        export PATH="$HOME/.cargo/bin:$PATH"
+        composer install --no-progress
+        vendor/bin/atoum --php 'php -d extension=wasm' --directories tests/units --force-terminal --use-dot-report
+  
+
 # List of all jobs.
 jobs:
-  # Build and test the project the project.
-  build_and_test:
+  # Build and test the project the project against the latest PHP version.
+  build_and_test_php_latest:
     docker:
       - image: circleci/php:latest
-    steps:
-      # Update the project.
-      - checkout
+    steps: *build_and_test_steps
 
-      # Install Rust.
-      - run:
-          name: Install Rust
-          command: |
-            test -d /usr/local/cargo || curl https://sh.rustup.rs -sSf | sh -s -- -y
-
-      # Install `just` used to manage the project.
-      - run:
-          name: Install just
-          command: |
-            export PATH="$HOME/.cargo/bin:$PATH"
-            test -f $HOME/.cargo/bin/just || cargo install just
-
-      # Install Composer.
-      - run:
-          name: Install Composer
-          command: |
-            php -r "copy('https://getcomposer.org/installer', 'composer-setup.php');"
-            php composer-setup.php
-            php -r "unlink('composer-setup.php');"
-            mv composer.phar composer
-
-      # Compile and install the PHP extension.
-      # `just build` isn't used because there is no some issue with
-      # `--with-pic` I don't understand…
-      - run:
-          name: Compile and install the PHP extension
-          command: |
-            export PATH="$HOME/.cargo/bin:$PATH"
-            export CXX="gcc"
-            EXTENSION="$HOME/project/src"
-            cd $EXTENSION
-            PHP_PREFIX=$(php-config --prefix)
-            PHP_PREFIX_BIN=$PHP_PREFIX/bin
-            $PHP_PREFIX_BIN/phpize --clean
-            $PHP_PREFIX_BIN/phpize
-            ./configure --with-php-config=$PHP_PREFIX_BIN/php-config
-            /bin/bash $EXTENSION/libtool --mode=compile $CXX -I. -I$EXTENSION -DPHP_ATOM_INC -I$EXTENSION/include -I$EXTENSION/main -I$EXTENSION -I$PHP_PREFIX/include/php -I$PHP_PREFIX/include/php/main -I$PHP_PREFIX/include/php/TSRM -I$PHP_PREFIX/include/php/Zend -I$PHP_PREFIX/include/php/ext -I$PHP_PREFIX/include/php/ext/date/lib -DHAVE_CONFIG_H -c $EXTENSION/wasm.cc -o wasm.lo -fPIC
-            $CXX -I. -I$EXTENSION -DPHP_ATOM_INC -I$EXTENSION/include -I$EXTENSION/main -I$EXTENSION -I$PHP_PREFIX/include/php -I$PHP_PREFIX/include/php/main -I$PHP_PREFIX/include/php/TSRM -I$PHP_PREFIX/include/php/Zend -I$PHP_PREFIX/include/php/ext -I$PHP_PREFIX/include/php/ext/date/lib -DHAVE_CONFIG_H -c $EXTENSION/wasm.cc  -DPIC -o .libs/wasm.o -fPIC
-            /bin/bash $EXTENSION/libtool --mode=link cc -DPHP_ATOM_INC -I$EXTENSION/include -I$EXTENSION/main -I$EXTENSION -I$PHP_PREFIX/include/php -I$PHP_PREFIX/include/php/main -I$PHP_PREFIX/include/php/TSRM -I$PHP_PREFIX/include/php/Zend -I$PHP_PREFIX/include/php/ext -I$PHP_PREFIX/include/php/ext/date/lib  -DHAVE_CONFIG_H  -g -O2    -o wasm.la -export-dynamic -avoid-version -prefer-pic -module -rpath $EXTENSION/modules  wasm.lo -Wl,-rpath,$EXTENSION/. -L$EXTENSION/. -lwasmer_runtime_c_api -fPIC
-            cc -shared  .libs/wasm.o  -L$EXTENSION/. -lwasmer_runtime_c_api  -Wl,-rpath -Wl,$EXTENSION/. -Wl,-soname -Wl,wasm.so -o .libs/wasm.so -fPIC
-            sudo make install-modules
-
-      # Run the extension test suites.
-      - run:
-          name: Test the extension
-          command: |
-            export PATH="$HOME/.cargo/bin:$PATH"
-            composer install --no-progress
-            vendor/bin/atoum --php 'php -d extension=wasm' --directories tests/units --force-terminal --use-dot-report
+  # Build and test the project the project against PHP 7.2.
+  build_and_test_php_7.2:
+    docker:
+      - image: circleci/php:7.2
+    steps: *build_and_test_steps
 
 
 # List of workflows.
@@ -69,8 +78,8 @@ workflows:
   # The build workflow.
   build:
     jobs:
-      # Run the `build_and_test` job for all branches and all tags.
-      - build_and_test:
+      # Run the `build_and_test_php_latest` job for all branches and all tags.
+      - build_and_test_php_latest: &build_and_test_php_latest
           filters:
             branches:
               only:
@@ -78,3 +87,6 @@ workflows:
                 - staging
             tags:
               only: /.*/
+
+      # Run the `build_and_test_php_7.2` job for all branches and all tags.
+      - build_and_test_php_7.2: *build_and_test_php_latest

--- a/.circleci/config.yml
+++ b/.circleci/config.yml
@@ -65,7 +65,7 @@ jobs:
     steps: *build_and_test_steps
 
   # Build and test the project the project against PHP 7.2.
-  build_and_test_php_7.2:
+  build_and_test_php_7_2:
     docker:
       - image: circleci/php:7.2
     steps: *build_and_test_steps
@@ -88,5 +88,5 @@ workflows:
             tags:
               only: /.*/
 
-      # Run the `build_and_test_php_7.2` job for all branches and all tags.
-      - build_and_test_php_7.2: *build_and_test_php_latest
+      # Run the `build_and_test_php_7_2` job for all branches and all tags.
+      - build_and_test_php_7_2: *build_and_test_php_latest

--- a/bors.toml
+++ b/bors.toml
@@ -1,6 +1,6 @@
 status = [
     "ci/circleci: build_and_test_php_latest",
-    "ci/circleci: build_and_test_php_7.2"
+    "ci/circleci: build_and_test_php_7_2"
 ]
 required_approvals = 0
 timeout_sec = 900

--- a/bors.toml
+++ b/bors.toml
@@ -1,4 +1,7 @@
-status = [ "ci/circleci: build_and_test" ]
+status = [
+    "ci/circleci: build_and_test_php_latest",
+    "ci/circleci: build_and_test_php_7.2"
+]
 required_approvals = 0
 timeout_sec = 900
 delete_merged_branches = true 

--- a/composer.json
+++ b/composer.json
@@ -16,7 +16,7 @@
         "chat": "https://spectrum.chat/wasmer"
     },
     "require": {
-        "php": "^7.3",
+        "php": "^7.2",
         "psr/simple-cache": "^1.0"
     },
     "require-dev": {

--- a/src/wasm.cc
+++ b/src/wasm.cc
@@ -239,7 +239,9 @@ public:
 
     ~wasm_lazy_byte_array_t()
     {
-        efree((uint8_t *) byte_array->bytes);
+        if (byte_array != NULL) {
+            efree((uint8_t *) byte_array->bytes);
+        }
     }
 
     wasmer_byte_array *get_bytes()

--- a/src/wasm.hh
+++ b/src/wasm.hh
@@ -48,7 +48,7 @@
  * `ZEND_ACC_IMMUTABLE` has been defined in PHP 7.3. Be compatible
  * with PHP 7.2 by defining an empty value.
  */
-#ifndef ZEND_ACC_IMMUTABLE
+#if !defined(ZEND_ACC_IMMUTABLE)
 # define ZEND_ACC_IMMUTABLE 0
 #endif
 
@@ -56,7 +56,7 @@
  * `ZEND_PARSE_PARAMETERS_NONE` has been defined in PHP 7.3. Be
  * compatible with PHP 7.2 by copy-pasting its definition.
  */
-#ifndef ZEND_PARSE_PARAMETERS_NONE
+#if !defined(ZEND_PARSE_PARAMETERS_NONE)
 # define ZEND_PARSE_PARAMETERS_NONE() \
     ZEND_PARSE_PARAMETERS_START(0, 0) \
     ZEND_PARSE_PARAMETERS_END()

--- a/src/wasm.hh
+++ b/src/wasm.hh
@@ -35,6 +35,14 @@
 #  include <stdint.h>
 #endif
 
+/**
+ * `ZEND_ACC_IMMUTABLE` has been defined in PHP 7.3. Be compatible
+ * with PHP 7.2 by defining an empty value.
+ */
+#ifndef ZEND_ACC_IMMUTABLE
+# define ZEND_ACC_IMMUTABLE 0
+#endif
+
 // Constant to represent a not nullable (return) type.
 #define NOT_NULLABLE 0
 

--- a/src/wasm.hh
+++ b/src/wasm.hh
@@ -43,6 +43,16 @@
 # define ZEND_ACC_IMMUTABLE 0
 #endif
 
+/**
+ * `ZEND_PARSE_PARAMETERS_NONE` has been defined in PHP 7.3. Be
+ * compatible with PHP 7.2 by copy-pasting its definition.
+ */
+#ifndef ZEND_PARSE_PARAMETERS_NONE
+# define ZEND_PARSE_PARAMETERS_NONE() \
+    ZEND_PARSE_PARAMETERS_START(0, 0) \
+    ZEND_PARSE_PARAMETERS_END()
+#endif
+
 // Constant to represent a not nullable (return) type.
 #define NOT_NULLABLE 0
 

--- a/src/wasm.hh
+++ b/src/wasm.hh
@@ -36,6 +36,15 @@
 #endif
 
 /**
+ * `_IS_NUBMER` has been defined in PHP 7.3. Be compatible with PHP
+ * 7.2 by defining the real value. It will have no effect, but the
+ * code will compile.
+ */
+#if !defined(_IS_NUMBER)
+# define _IS_NUMBER 20
+#endif
+
+/**
  * `ZEND_ACC_IMMUTABLE` has been defined in PHP 7.3. Be compatible
  * with PHP 7.2 by defining an empty value.
  */

--- a/src/wasm.hh
+++ b/src/wasm.hh
@@ -62,6 +62,25 @@
     ZEND_PARSE_PARAMETERS_END()
 #endif
 
+/**
+ * `zend_register_persistent_resource_ex` has been added in PHP 7.3
+ * (see
+ * https://github.com/php/php-src/commit/67d5f39a47b15e28293d9d6558b80ded049179fe). Be
+ * compatible with PHP 7.2 by writing a small hack that should work.
+ */
+#if PHP_VERSION_ID < 70300
+ZEND_API zend_resource* zend_register_persistent_resource_ex(zend_string *key, void *rsrc_pointer, int rsrc_type)
+{
+    zval *zv;
+    zval tmp;
+
+    ZVAL_NEW_PERSISTENT_RES(&tmp, -1, rsrc_pointer, rsrc_type);
+    zv = zend_hash_update(&EG(persistent_list), key, &tmp);
+
+    return Z_RES_P(zv);
+}
+#endif
+
 // Constant to represent a not nullable (return) type.
 #define NOT_NULLABLE 0
 

--- a/tests/units/Extension/Classes.php
+++ b/tests/units/Extension/Classes.php
@@ -258,7 +258,7 @@ class Classes extends Suite
                 ->let($return_type = $methods[3]->getReturnType())
 
                 ->string($return_type . '')
-                    ->isEqualTo('number')
+                    ->isEqualTo(PHP_VERSION_ID < 70300 ? 'unknown' : 'number')
                 ->boolean($return_type->allowsNull())
                     ->isFalse()
 
@@ -412,7 +412,7 @@ class Classes extends Suite
         $this
             ->given(
                 $byteLength = 1114112,
-                $wasmArrayBuffer = new WasmArrayBuffer($byteLength),
+                $wasmArrayBuffer = new WasmArrayBuffer($byteLength)
             )
             ->when($result = $wasmArrayBuffer->grow(1))
             ->then
@@ -655,7 +655,7 @@ class Classes extends Suite
             ->given(
                 $wasmArrayBuffer = new WasmArrayBuffer($wasmTypedArrayClassName::BYTES_PER_ELEMENT),
                 $wasmTypedArray = new $wasmTypedArrayClassName($wasmArrayBuffer),
-                $wasmTypedArray[0] = 42,
+                $wasmTypedArray[0] = 42
             )
             ->when($result = $wasmTypedArray[0])
             ->then

--- a/tests/units/Extension/Functions.php
+++ b/tests/units/Extension/Functions.php
@@ -282,7 +282,7 @@ class Functions extends Suite
                 ->let($return_type = $_result->getReturnType())
 
                 ->string($return_type . '')
-                    ->isEqualTo('number')
+                    ->isEqualTo(PHP_VERSION_ID < 70300 ? 'unknown' : 'number')
                 ->boolean($return_type->allowsNull())
                     ->isTrue()
 

